### DISCUSSION TITLE
i2c on Spike Prime

### DISCUSSION THREAD

**Discussion by dmusican** [pybricks/support#1047] [Q&A] [answered]
I've got a number of HiTechnic IR Seekers, which I've got left over from my days of using LEGO NXT devices. I'd love to make them work on my Spike Prime bots. Those HiTechnic sensors use I2C, so in principle, if I can get I2C going on the Spike Prime, I should be able to make it work. I've already got an adapter cable that I wrangled up that conects the Powered Up connector on the SpikePrime and the old LEGO style connectors. (I can't guarantee that the cable is wired correctly, but I'm optimistic; I won't know for sure without trying to make it work on the software side.)

I know that pybricks is capable of [I2C communication on the EV3](https://docs.pybricks.com/en/v3.0/iodevices/i2cdevice.html). That page I just linked says

> This class is only supported on the EV3 at this time. It could be added to Powered Up hubs in a future release. If you’d like to see this happen, be sure to ask us on our [support page](https://github.com/pybricks/support/issues/).

So, I'm asking! Can it be added? Alternatively, is there some other sort of hack I should be looking at? I could try to write my own I2C implementation from scratch via bit-banging, but it would be lovely to not have to do that, especially if this might be coming via pybricks in the future.

**Comment by laurensvalk** [comment, marked as answer]
It's not out of the question, but it will need quite a bit of work if we want to do it properly. See https://github.com/pybricks/pybricks-micropython/pull/159 for some very early ideas about making the ports multi-functional.

A small number of ports on SPIKE may support I2C via hardware, but the more likely route may be a software based solution so that it could work on all hubs/ports. This is essentially still bit-banging, but with a lower level process so it doesn't depend on the timing of Python code.

**Reply by dmusican** [reply]
Thanks so much for the quick and helpful response. I'm going to take this as meaning that this feature is not coming anytime soon, though of course it would be welcome if it did.

I'm trying to decide whether to choose to spend money instead of time and just buy some [BBS IR Seekers](https://buildingblockstudio.com/product/spike-ir-board/); that might be the more expedient route.

**Comment by mihaitutu** [comment]
I know that this has been addressed, yet I would really like to know if we can interface other sensors via I2C now. Has this been worked on lately?
 
For example, this year in WRO, we can use any kind of platforms and sensors. We want to stay competitive and having such an option might become more interesting for students in Senior category.

**Reply by laurensvalk** [reply]
Not any time soon, but we have made some incremental steps here. We are working on a [new firmware for EV3](https://pybricks.com/project/saving-lego-mindstorms/). It is going to need a software based I2C driver, likely based on the ev3dev FIQ driver. We aim to generalize most code to work on multiple platforms, so maybe it can work on SPIKE too.

But there may be a better solution that is also closer to coming to fruition. We are also likely enabling direct UART access, so you could wire up just about anything or any other board.

Also @dmusican -- going back to your original question:

> I've got a number of HiTechnic IR Seekers, which I've got left over from my days of using LEGO NXT devices. I'd love to make them work on my Spike Prime bot

If we had EV3/NXT support in Pybricks (browser-based and fast like SPIKE Prime), would you still use it?

**Reply by dmusican** [reply]
Thanks! Not any more... I got rid of my NXT bricks when I replaced them with the SPIKE Primes. Following your question, I'm now regretting having done so!

**Comment by kai-morich** [comment]
Instead of connecting devices directly to the Spike I am now experimenting with the LMS-ESP32-v2 board from https://www.antonsmindstorms.com/.
I already attached some devices and collected some code snippets at https://gist.github.com/kai-morich as preparation for the new WRO season.
